vignettes/articles/managing-tokens-securely.Rmd

```diff
@@ -0,0 +1,198 @@
+
+```{r setup}
+library(gargle)
+```
+
+## Overview of the approach
+
+  * Generate a random PASSWORD and give it a self-documenting name, e.g. `GARGLE_PASSWORD`.
+    Store as an environment variable.
+  * Identify a secret file of interest, such as the JSON representing a
+    service account token. This is presumably stored *outside* your package.
+  * Use the PASSWORD to apply a method for symmetric encryption to the target
+    file.
+  * Store the encrypted file in a predictable place *within* your package.
+  * Store or pass the PASSWORD as an environment variable everywhere you'll need
+    to decrypt the secret.
+    - Check that the platform has support for keeping the PASSWORD concealed.
+    - Make sure you don't do anything in your own code that would dump it to,
+      e.g., a log file.
+  * Rig your tests to determine if the key is available and, therefore,
+    whether decryption is going to be possible.
+    - If "no", carry on gracefully with any tests that don't require auth.
+    - If "yes", decrypt the secret and put the associated token into force
+      globally for the test run or on an "as needed" basis in individual tests.
+
+## Annotated code-through
+
+### Generate a name for the PASSWORD
+
+We believe this should record the associated package. `secret_pw_name()` does this.
+
+```{r}
+(pw_name <- gargle:::secret_pw_name("gargle"))
+```
+
+### Generate a random PASSWORD
+
+In real life, you should keep the output of `secret_pw_gen()` to yourself! We reveal it here as part of the exposition.
+
+```{r}
+(pw <- gargle:::secret_pw_gen())
+```
+
+### Define env var in `.Renviron`
+
+Combine the name and value to form a line like this in your user-level `.Renviron` file:
+
+```{r, echo = FALSE, results = "asis"}
+cat(paste0(pw_name, "=", pw), sep = "\n")
+```
+
+`usethis::edit_r_environ()` can help create or open this file. We recommend using the user-level `.Renviron`, as opposed to project-level, for the usual reason  that this makes it less likely you will share sensitive information by mistake. Make sure `.Renviron` ends in a newline; the lack of this is a notorious cause of silent failure. Remember you'll need to restart R for the newly defined environment variable to take effect.
+
+### Provide env var to other services
+
+On Travis-CI, define this env var in your repo settings via the browser UI:
+
+https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings
+
+Alternatively, you can use the Travis command line interface to configure the env var or even define an encrypted env var in `.travis.yml`. In any case, remember that private env vars are not available to pull requests, which is another reason to carry on gracefully when token decryption is not possible (see below).
+
+On R-hub, send this env var in your calls to `check()` and friends:
+
+```
+rhub::check(env_vars = Sys.getenv(secret_pw_name("gargle"), names = TRUE))
```

Comment:
should this be `gargle:::secret_pw_name`?

Reply:
Yeah this one should. Thanks.